### PR TITLE
Updated Database Schema

### PR DESCRIPTION
**Pull request by chrisweilacker**
-Refactored MySQL Code and Small Amount of Front-End Code
-Created 10M Record Seed Script
-Modularized DB Calls to allow easy switching between several DBs

### CODE REVIEW

File: database/mysql.js [hunk 1]
```diff
@@ -1,28 +1,31 @@
...
-const interiorTable = 'CREATE TABLE IF NOT EXISTS interior_features (feature_id int NOT NULL AUTO_INCREMENT, bedrooms varchar(255), bathrooms varchar (255), interiorheating varchar (255), interiorcooling varchar (255), appliances varchar (255), kitchen varchar (255), flooring varchar (255), house_id varchar (255), sqft varchar (255), PRIMARY KEY (feature_id));'
+const featuresTable = 'CREATE TABLE IF NOT EXISTS features (house_id int NOT NULL AUTO_INCREMENT, type varchar(255), year_built int, heating varchar (255), cooling varchar (255), parking varchar (255),lot int, days_on_zillow int, bedrooms float, bathrooms float, interiorheating varchar (255), interiorcooling varchar (255), appliances varchar (255), kitchen varchar (255), flooring varchar (255), sqft int, PRIMARY KEY (house_id));'
 const dbName = 'zillow';
 const mockFeatures = require('./mockFeatures');
 const mockInterior = require('./mockInterior');
 const features = require('./createMockData');
```
Comment (ckallemeyn) [inline database/mysql.js:6]: This is a nit pick suggestion, but may help readability. On Line 7 since you are only importing the createFeatures function from createMockData.js why not specific the function in your require statement. Here is an example:

`const createFeatures = require('./createMockData').createFeatures;`

The other easy fix would be to just export the function instead of putting it in an object on line 73 of  your createMockData file. Here's what I mean:

`module.exports = createFeatures;`

File: database/mysql.js [hunk 2]
```diff
@@ -35,55 +38,38 @@ const fs = require('fs');
...
-        db.query(sql, params, (err, results) => {
-          if (err) {
-            console.log(err);
-          } else {
-            console.log("mockFeatures loaded", results);
-          }
-        });
+    let sql = 'INSERT into features (type, year_built, heating, cooling, parking, lot, days_on_zillow, bedrooms, bathrooms, interiorheating, interiorcooling, appliances, kitchen, flooring, sqft) VALUES ?';
+    let params = [];
+    features.createFeatures().forEach(house => {
```
Comment (ckallemeyn) [inline database/mysql.js]: Another nit pick... mainly about syntax style standards. On line 57, I see that you use the "house" parameter without any "()" surrounding the param, but for most of your other anonymous arrow functions you are using parens. It might help for style consistency to use the parens for all arrow functions. Again, this is a major nit pick.

File: client/src/components/index.jsx [hunk 3]
```diff
@@ -33,51 +33,42 @@ class App extends React.Component {
...
-            appliances: house.appliances,
-            flooring: house.flooring +' sqft'
-          });
-        })
-        .catch( error => {
-          console.log(error);
-        });
-
   }
   render() {
```
Comment (ckallemeyn) [inline client/src/components/index.jsx:63]: Great job with the refactor. This is a suggestion that might help with decreasing character count. Inside the render function you can deconstruct the "this.state" object properties and then reference them inside the "Facts_Features" and "InteriorFeatures" components. Here's an example: 

line 63 1/2 `const { type, yearBuilt, factsCooling, factsHeating, parking, lot} = this.state;`

line 67 ```
<Facts_Features type={type} yearBuilt={yearBuilt} factsCooling={factsCooling}
            factsHeating={factsHeating} parking={parking} lot={lot}/>
```

Again, this is just a suggestion it does not improve performance, but it does "DRY" up the code.

File: database/createMockData.js [hunk 4]
```diff
@@ -30,48 +18,41 @@ const get_random = function (array) {
...
+         appliances: get_random(houseFeatures.appliances),
+         int_heating: get_random(houseFeatures.intHeating),
+         int_cooling: get_random(houseFeatures.intCooling),
+         kitchen:  get_random(houseFeatures.kitchen_features),
+         flooring: houseFeatures.flooring(1000,3000),
+         sqft: houseFeatures.flooring(1000,3000),
+         days_on_zillow: houseFeatures.flooring(1, 100)
        });
      }
    return features;
```
Comment (ckallemeyn) [inline database/createMockData.js:51]: Great job here! I just noticed some small spacing issues on the properties that do not use the "get_random" func. I would just separate the two param numbers by a space after each coma. Example: 

`line 45: flooring: houseFeatures.flooring(1000, 3000)`

**Review by ckallemeyn — CHANGES_REQUESTED** [review]

File: database/mysql.js (hunk 1 above)
Comment (chrisweilacker) [reply database/mysql.js:6]: True, left over from my refactor where I deleted the interiorFeatures function.

File: database/mysql.js (hunk 2 above)
Comment (chrisweilacker) [reply database/mysql.js]: Yea, I copied this from a piece of Clifton's original code. You can tell as I don't think I would have chained the forEach like that normally.  Usually I would set it into its own variable and then called the forEach.

File: client/src/components/index.jsx (hunk 3 above)
Comment (chrisweilacker) [reply client/src/components/index.jsx:63]: Great idea to look into, still not used to doing destructing normally, but should get used to it.

File: database/createMockData.js (hunk 4 above)
Comment (chrisweilacker) [reply database/createMockData.js:51]: Gotcha, this is why one should hook up their linter.

**Comment by LesliePajuelo** [comment]
Please remove the bundle file from the commit

File: database/create10MRecords.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+//How many records to randomly create per batch insert
+const recordsPerInsert = 100000;
+//The number of simultaneous function instances to be called,
+// I found 5 to be optimal if using generated data on each insert and 2 if using static data, for Mongo we must use one.
+const instances = (process.env.DB.trim() === 'mongo') ? 1 : 2;
+const numberOfLoops= Math.ceil(recordsToInsert/recordsPerInsert);
+//Tested with static Data and found not much improvement over randomizing each insert
+//probably due to the multiple simultaneous function calls becoming less productive.
+var staticData;
+
```
Comment (LesliePajuelo) [inline database/create10MRecords.js:17]: It's best to avoid variables, especially ones with a large amount of data on a higher scope than necessary.

Any time this file is open it will prevent staticData from being garbage collected. Only create this variable right when you use it.

File: database/create10MRecords.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+db.createConnection(function () {
+  staticData = db.createParams(recordsPerInsert);
+  //set the start time for timing the insert of Records
+  start = new Date().getTime();
+  //start several simultaneous function calls
```
Comment (LesliePajuelo) [inline database/create10MRecords.js:22]: Counting time from when the connection starts will artificially inflate your performance times.

File: database/create10MRecords.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+//The number of simultaneous function instances to be called,
+// I found 5 to be optimal if using generated data on each insert and 2 if using static data, for Mongo we must use one.
+const instances = (process.env.DB.trim() === 'mongo') ? 1 : 2;
+const numberOfLoops= Math.ceil(recordsToInsert/recordsPerInsert);
+//Tested with static Data and found not much improvement over randomizing each insert
+//probably due to the multiple simultaneous function calls becoming less productive.
+var staticData;
+
+db.createConnection(function () {
+  staticData = db.createParams(recordsPerInsert);
```
Comment (LesliePajuelo) [inline database/create10MRecords.js:19]: This function call,  staticData = db.createParams(recordsPerInsert); should be right before the data is used

File: database/create10MRecords.js (hunk 5 above)
Comment (LesliePajuelo) [reply database/create10MRecords.js:17]: Most of these variables should be closer to where they are used.

File: database/create10MRecords.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  //set the start time for timing the insert of Records
+  start = new Date().getTime();
+  //start several simultaneous function calls
+  for (var i = 0; i < instances; i++) {
+    nextStep();
+  }
+});
+
+function nextStep() {
```
Comment (LesliePajuelo) [inline database/create10MRecords.js:28]: It looks like some of the variables are used in nextStep and load10M... Send it through as an argument.

load10MRecords(recordsPer....)

File: database/create10MRecords.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+          if (err) {
+            reject(err);
+          } else {
+            resolve('loop');
+          } //finally
+        }, completedInserts === numberOfLoops - 1 ? (recordsToInsert - (completedInserts * recordsPerInsert)) : recordsPerInsert,
+        staticData);
+      });
+      let result = promise;
+      return result;
```
Comment (LesliePajuelo) [inline database/create10MRecords.js:57]: Why not just return promise?

Or better yet return on line 46?

File: database/createMockData.js [hunk 10]
```diff
@@ -1,25 +1,13 @@
 
-const get_random = function (array) {
+var get_random = function (array) {
    if (array.length < 1) {
```
Comment (LesliePajuelo) [inline database/createMockData.js:3]: Why the change from const to var?

File: database/index.js [hunk 11]
```diff
@@ -1,11 +1,15 @@
-if (!process.env.DB) {process.env.DB = 'mysql'}
+if (!process.env.DB) {process.env.DB = 'postgres'}
```
Comment (LesliePajuelo) [inline database/index.js:1]: This is a great use case for a ternary

File: database/mongoDb.js [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+      appliances: String,
+      kitchen: String,
+      flooring: String,
+      sqft: Number
+    }, { _id : false });
+    featuresTable = mongoose.model('features', schema);
+    console.log('Connected');
+    featuresTable.countDocuments({}, function(err, count) {
+      if (err) { return handleError(err) } //handle possible errors
+
```
Comment (LesliePajuelo) [inline database/mongoDb.js:43]: handleError is not in this review, so not sure what it handles. This is a great use case for having a top level catch and throw in function

mdn.io/catchthrow

File: database/mongoDb.js [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+      documentCount = count;
+      console.log('The amount of records is: ' + documentCount);
+      cb();
```
Comment (LesliePajuelo) [inline database/mongoDb.js:46]: cb? what does the cb do? Meaningful name please

File: database/mongoDb.js [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+              flooring: house.flooring,
+              sqft: house.sqft
+            }
+          }
+        });
+      });
+    } else {
+      params.forEach(function (value, index, array) {
+        value.insertOne.document.house_id = ++documentCount;
+      });
```
Comment (LesliePajuelo) [inline database/mongoDb.js:119]: What is the case that params will already have values at this point?

File: database/mongoDb.js [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+      } else {
+        console.log(docs);
+        callback(null, docs);
+      }
+    });
+    };
+
+    var getAllFeatures = (page, callback) => {
+
+    };
```
Comment (LesliePajuelo) [inline database/mongoDb.js:161]: You're going to want params for which records page you're on. So a min, max.

Very well built APIs will also allow for some options, such as ascending/descending order. No need to implement those now, but especially for getAll type endpoints it'll be added later

so getAllFeatures = (page, options, callback)
Where options is {min, max, order, date....etc}

File: database/mongoDb.js [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+
+    var postFeatures = (features, callback) => {
+      if (isNaN(id)) {
+        callback('no Id to insert', null);
+      }
+      features.house_id = undefined;
```
Comment (LesliePajuelo) [inline database/mongoDb.js:167]: undefined is a value set by the interpreter. Use null.

File: database/mongoDb.js [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+      }
+      features.house_id = id;
+      features = keepOnlyFieldNames(features);
+    };
+
+    var deleteFeatures = (id, callback) => {
+
+    };
+
+    var keepOnlyFieldNames = ({ house_id, type, year_built, heating, cooling, parking, lot, days_on_zillow, bedrooms, bathrooms, interiorheating, interiorcooling, appliances, kitchen, flooring, sqft}) => {
```
Comment (LesliePajuelo) [inline database/mongoDb.js:184]: Why? Why is this happening?

File: database/mysql.js [hunk 18]
```diff
@@ -123,11 +104,59 @@ const fs = require('fs');
...
+    var getAllFeatures = (page, callback) => {
+
+    };
+
+    var postFeatures = (features, callback) => {
+      if (isNaN(id)) {
+        callback('no Id to insert', null);
+      }
+      features.house_id = undefined;
+      features = keepOnlyFieldNames(features);
```
Comment (LesliePajuelo) [inline database/mysql.js:116]: features = Object.keys(features)

Why do you need features to be an object instead of an array?

if you really need it to be features = {... Object.keys(features)} Or features = (features) =>{
return features.each...}

Or use lodash _.forKey, you can import just that function. Hard coding field names/features, eek!

File: database/mysql.js [hunk 19]
```diff
@@ -123,11 +104,59 @@ const fs = require('fs');
...
+      features.house_id = id;
+      features = keepOnlyFieldNames(features);
+    };
+
+    var deleteFeatures = (id, callback) => {
+
+    };
+
+    var keepOnlyFieldNames = ({ house_id, type, year_built, heating, cooling, parking, lot, days_on_zillow, bedrooms, bathrooms, interiorheating, interiorcooling, appliances, kitchen, flooring, sqft}) => {
+      return ({
```
Comment (LesliePajuelo) [inline database/mysql.js:133]: See comment where invoked.

File: database/postgres.js [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+    });
+
+    return params;
+  }
+
+
+  //function to load feature mock data
+  var loadFeatures = (callback, numRecords = 100, params = []) => {
+    // fs.writeFileSync('/Users/apple/Code/RPT11/zillowclone/database/mockfeature.json', JSON.stringify(features.createFeatures()));
+
```
Comment (LesliePajuelo) [inline database/postgres.js:132]: Good call, avoid Sync methods

File: database/postgres.js [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+    if (params.length === 0) {
+      features.createFeatures(numRecords).forEach(house => {
+        params.push([house.type, house.year_built, house.heating, house.cooling, house.parking, house.lot, house.days_on_zillow, house.bedrooms, house.bathrooms, house.int_heating, house.int_cooling, house.appliances, house.kitchen, house.flooring, house.sqft]);
+      });
```
Comment (LesliePajuelo) [inline database/postgres.js:136]: Why not use the createParams function above?

File: database/postgres.js [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+        house_id: id
+      }
+    }).then((results) => {
+        callback(null, results);
+    }).catch((err) => {
+      callback(err, null);
+    });
+  };
+
+  var keepOnlyFieldNames = ({ house_id, type, year_built, heating, cooling, parking, lot, days_on_zillow, bedrooms, bathrooms, interiorheating, interiorcooling, appliances, kitchen, flooring, sqft}) => {
```
Comment (LesliePajuelo) [inline database/postgres.js:230]: Other than not using this function, have you thought of a utils file?

File: database/mongoDb.js (hunk 13 above)
Comment (LesliePajuelo) [reply database/mongoDb.js:46]: The cb is an empty function.

Why does this exist?

File: test/testspec.js [hunk 23]
```diff
@@ -22,21 +22,13 @@ describe('Testing Database', () => {
...
-    });
   });
 
   it("should have 100 records in the features table", function () {
     connection.query('SELECT * from features', (err, results) => {
       console.log(results)
-        expect(results).to.equal(100);
+        expect(results).to.be.greaterthan(100);
         done();
     });
```
Comment (LesliePajuelo) [inline test/testspec.js:33]: I see you just modified this test. However, it's testing your database not your javascript. 
OK if it's your first tests and it bails out of the test suite on failure to prevent the rest of your tests from running

**Review by LesliePajuelo — COMMENTED** [review]
Overall good. Some more work to reduce specificity and future proofing.

File: database/mongoDb.js (hunk 17 above)
Comment (chrisweilacker) [reply database/mongoDb.js:184]: Wanted to make sure if the user/front-end sends in something misspelled or something random will remove these fields. Probably was overthinking it because mongoose does have a schema that would be violated and return an error rather than straight mongo which I believe can accept anything. Actually at the time thought this was a smart way only had to write the key names once then copy and paste.

File: database/mongoDb.js (hunk 16 above)
Comment (chrisweilacker) [reply database/mongoDb.js:167]: If I do that won't mongoose try to insert null into house_id for this record and error out?  I used undefined to make the key effectively 'dissappear' at least that is what went through my mind.

File: database/postgres.js (hunk 21 above)
Comment (chrisweilacker) [reply database/postgres.js:136]: Totally should do that.  I created the createParams function to have staticData for the 10M records this was basically the original code from FEC which only runs when 100 records is invoked.

File: database/mongoDb.js (hunk 14 above)
Comment (chrisweilacker) [reply database/mongoDb.js:119]: Inserting the 31250 records of static Data from 10M Records seed script, otherwise will be created when called from the 100 records seed script which I plan to use for travisCI testing.

File: database/mongoDb.js (hunk 13 above)
Comment (chrisweilacker) [reply database/mongoDb.js:46]: Was used when I needed something to happen after the DB was fully opened.  I do have a function sent to it in the 10M DB script where I need to wait to get the amount of records in the DB for Mongo before I start inserting in order to do my own version of an Auto Increment for the House_id field.  Auto incrementing in Mongo was something I had issues with.

File: test/testspec.js (hunk 23 above)
Comment (chrisweilacker) [reply test/testspec.js:33]: Writing more tests now.

File: database/postgres.js (hunk 20 above)
Comment (chrisweilacker) [reply database/postgres.js:132]: Cant take credit was commented before in the FEC code and should probably just delete it.

**Review by chrisweilacker — COMMENTED** [review]
Thanks for the feedback.
